RPC Error leaving venerable file behind

On two of our projects, we have two different deployments into two different cloud-foundry spaces. ( totally independent of one another  ) we occasionally receive an RCP error. The RPC error then leaves a copy of the {application}-venerable behind, and the next deployment fails with unable to rename.

We deploy a docker image using Blue-green strategy:

```
resource "cloudfoundry_app" "app_application" {
    name         = var.paas_app_application_name
    space        = data.cloudfoundry_space.space.id
    docker_image = 'dfedigital/get-into-teaching-frontend:xxxxx' 
    strategy     = 'blue-green'
    memory       = 1024
    dynamic "service_binding" {
      for_each = data.cloudfoundry_user_provided_service.logging
      content {
        service_instance = service_binding.value["id"]
      }
    }

    service_binding  {
            service_instance = cloudfoundry_service_instance.redis.id
    }

    routes { route = cloudfoundry_route.app_route_internal.id }
    dynamic "routes" {
      for_each = data.cloudfoundry_route.app_route_internet
      content {
        route = routes.value["id"]
      }
    }
    environment = {
           REDACTED
    }
}
```

The first error we see is :
```
2020-09-16T12:59:53.5402571Z Acquiring state lock. This may take a few moments...
2020-09-16T12:59:56.5157282Z [0m[1mcloudfoundry_app.app_application: Modifying... [id=98341e16-a9e8-4145-a616-944ceae9ca66][0m[0m
2020-09-16T13:00:06.5162640Z [0m[1mcloudfoundry_app.app_application: Still modifying... [id=98341e16-a9e8-4145-a616-944ceae9ca66, 10s elapsed][0m[0m
2020-09-16T13:00:16.5163191Z [0m[1mcloudfoundry_app.app_application: Still modifying... [id=98341e16-a9e8-4145-a616-944ceae9ca66, 20s elapsed][0m[0m
2020-09-16T13:00:26.5167126Z [0m[1mcloudfoundry_app.app_application: Still modifying... [id=98341e16-a9e8-4145-a616-944ceae9ca66, 30s elapsed][0m[0m
2020-09-16T13:00:36.5169889Z [0m[1mcloudfoundry_app.app_application: Still modifying... [id=98341e16-a9e8-4145-a616-944ceae9ca66, 40s elapsed][0m[0m
2020-09-16T13:00:47.2564896Z [0m[1mcloudfoundry_app.app_application: Still modifying... [id=98341e16-a9e8-4145-a616-944ceae9ca66, 50s elapsed][0m[0m
2020-09-16T13:00:56.5171045Z [0m[1mcloudfoundry_app.app_application: Still modifying... [id=98341e16-a9e8-4145-a616-944ceae9ca66, 1m0s elapsed][0m[0m
2020-09-16T13:01:06.5172704Z [0m[1mcloudfoundry_app.app_application: Still modifying... [id=98341e16-a9e8-4145-a616-944ceae9ca66, 1m10s elapsed][0m[0m
2020-09-16T13:01:16.5173479Z [0m[1mcloudfoundry_app.app_application: Still modifying... [id=98341e16-a9e8-4145-a616-944ceae9ca66, 1m20s elapsed][0m[0m
2020-09-16T13:01:17.3971991Z [31m
2020-09-16T13:01:17.3973296Z [1m[31mError: [0m[0m[1mrpc error: code = Unavailable desc = transport is closing[0m
2020-09-16T13:01:18.7637852Z 
2020-09-16T13:01:18.7640093Z [0m[0m[0m
```

[logs_1164.zip](https://github.com/cloudfoundry-community/terraform-provider-cloudfoundry/files/5237606/logs_1164.zip)



Then after that we start to see the following errors on subsequent deployments:
```
2020-09-16T13:11:00.9178179Z [1m[31mError: [0m[0m[1mOh no. Something's gone wrong. I've tried to roll back but you should check to see if everything is OK.: The app name is taken: ***[0m
2020-09-16T13:11:00.9178910Z 
2020-09-16T13:11:00.9179938Z [0m  on application.tf line 1, in resource "cloudfoundry_app" "app_application":
2020-09-16T13:11:00.9181862Z    1: resource "cloudfoundry_app" "app_application" [4m{[0m
2020-09-16T13:11:00.9182525Z [0m
2020-09-16T13:11:00.9183276Z [0m[0m
```

[logs_1165.zip](https://github.com/cloudfoundry-community/terraform-provider-cloudfoundry/files/5237631/logs_1165.zip)

If we delete the application, then rename venerable back to the application name, we can keep the system available and subsequent deployments start to work.

The initial problem seems to be a RPC error, which I cannot trace. but could it be possible to retry if that happens.

Subsequently we have the <application>-venerable left, can that be deleted?

Here is the error (For crash.log and trace, run the above):
```
$ terraform apply -auto-approve                                                                                                                                     
data.cloudfoundry_domain.domain: Refreshing state... [id=86484fb4-8d55-4f24-a3af-eda030b95ad8]
data.cloudfoundry_org.org: Refreshing state... [id=386a9502-d9b6-4aba-b3c3-ebe4fa3f963e]
data.cloudfoundry_space.space: Refreshing state... [id=934ff889-4644-46b3-afb2-79f47c558c73]
cloudfoundry_route.web_app_route: Refreshing state... [id=e165ffa5-44ea-4c56-b9c1-a274ecfc01ef]
cloudfoundry_app.web_app: Refreshing state... [id=0d2fa87d-f647-4943-b905-6f7315e55e57]
cloudfoundry_app.web_app: Modifying... [id=0d2fa87d-f647-4943-b905-6f7315e55e57]
cloudfoundry_app.web_app: Still modifying... [id=0d2fa87d-f647-4943-b905-6f7315e55e57, 10s elapsed]

Error: rpc error: code = Unavailable desc = transport is closing


panic: runtime error: index out of range [0] with length 0
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: goroutine 30 [running]:
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: github.com/terraform-providers/terraform-provider-cloudfoundry/cloudfoundry/managers/appdeployers.BlueGreenV2.Restage.func5(0xc00085a150, 0xc00085a150, 0x0, 0x0)
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 	github.com/terraform-providers/terraform-provider-cloudfoundry/cloudfoundry/managers/appdeployers/bluegreenv2_strategy.go:152 +0x428
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: github.com/terraform-providers/terraform-provider-cloudfoundry/cloudfoundry/managers/appdeployers.Actions.Execute(0xc000869908, 0x5, 0x5, 0x2befbe0, 0x0, 0x0)
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 	github.com/terraform-providers/terraform-provider-cloudfoundry/cloudfoundry/managers/appdeployers/rewind.go:219 +0x86
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: github.com/terraform-providers/terraform-provider-cloudfoundry/cloudfoundry/managers/appdeployers.BlueGreenV2.Restage(0xc0004aa4a0, 0xc0005fc000, 0xc000466100, 0xc000488200, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 	github.com/terraform-providers/terraform-provider-cloudfoundry/cloudfoundry/managers/appdeployers/bluegreenv2_strategy.go:166 +0x3e8
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: github.com/terraform-providers/terraform-provider-cloudfoundry/cloudfoundry.resourceAppUpdate(0xc0001dd180, 0x1e90b00, 0xc0005fac80, 0x0, 0x0)
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 	github.com/terraform-providers/terraform-provider-cloudfoundry/cloudfoundry/resource_cf_app.go:540 +0xa3e
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc00043de60, 0xc000497310, 0xc000297400, 0x1e90b00, 0xc0005fac80, 0x1deb801, 0xc0006bb890, 0xc00082c240)
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 	github.com/hashicorp/terraform-plugin-sdk@v1.15.0/helper/schema/resource.go:316 +0x263
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc00010c000, 0xc0006bba10, 0xc000497310, 0xc000297400, 0xc000818088, 0xc00000e708, 0x1dee380)
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 	github.com/hashicorp/terraform-plugin-sdk@v1.15.0/helper/schema/provider.go:294 +0x99
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc00000e4d0, 0x22774e0, 0xc000203d40, 0xc0001dc230, 0xc00000e4d0, 0xc000203d40, 0xc0001fbb78)
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 	github.com/hashicorp/terraform-plugin-sdk@v1.15.0/internal/helper/plugin/grpc_provider.go:885 +0x8b4
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x1f571a0, 0xc00000e4d0, 0x22774e0, 0xc000203d40, 0xc00044eea0, 0x0, 0x22774e0, 0xc000203d40, 0xc000351500, 0x64f)
2020-10-05T18:02:46.392+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 	github.com/hashicorp/terraform-plugin-sdk@v1.15.0/internal/tfplugin5/tfplugin5.pb.go:3305 +0x217
2020-10-05T18:02:46.393+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: google.golang.org/grpc.(*Server).processUnaryRPC(0xc000682780, 0x2285880, 0xc000683800, 0xc000708400, 0xc000202a80, 0x2ba7e80, 0x0, 0x0, 0x0)
2020-10-05T18:02:46.393+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 	google.golang.org/grpc@v1.27.1/server.go:1024 +0x501
2020-10-05T18:02:46.393+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: google.golang.org/grpc.(*Server).handleStream(0xc000682780, 0x2285880, 0xc000683800, 0xc000708400, 0x0)
2020-10-05T18:02:46.393+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 	google.golang.org/grpc@v1.27.1/server.go:1313 +0xd3d
2020-10-05T18:02:46.393+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc00020a570, 0xc000682780, 0x2285880, 0xc000683800, 0xc000708400)
2020-10-05T18:02:46.393+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 	google.golang.org/grpc@v1.27.1/server.go:722 +0xa1
2020-10-05T18:02:46.393+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: created by google.golang.org/grpc.(*Server).serveStreams.func1
2020-10-05T18:02:46.393+0100 [DEBUG] plugin.terraform-provider-cloudfoundry_v0.12.4: 	google.golang.org/grpc@v1.27.1/server.go:720 +0xa1
2020-10-05T18:02:46.395+0100 [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2020/10/05 18:02:46 [DEBUG] cloudfoundry_app.web_app: apply errored, but we're indicating that via the Error pointer rather than returning it: rpc error: code = Unavailable desc = transport is closing
2020/10/05 18:02:46 [TRACE] eval: *terraform.EvalMaybeTainted
2020/10/05 18:02:46 [TRACE] eval: *terraform.EvalWriteState
2020/10/05 18:02:46 [TRACE] EvalWriteState: recording 2 dependencies for cloudfoundry_app.web_app
2020/10/05 18:02:46 [TRACE] EvalWriteState: writing current state object for cloudfoundry_app.web_app
2020-10-05T18:02:46.395+0100 [DEBUG] plugin: plugin process exited: path=.terraform/plugins/registry.terraform.io/cloudfoundry-community/cloudfoundry/0.12.4/darwin_amd64/terraform-provider-cloudfoundry_v0.12.4 pid=8940 error="exit status 2"
2020/10/05 18:02:46 [TRACE] eval: *terraform.EvalApplyProvisioners
2020/10/05 18:02:46 [TRACE] EvalApplyProvisioners: cloudfoundry_app.web_app is not freshly-created, so no provisioning is required
2020/10/05 18:02:46 [TRACE] eval: *terraform.EvalMaybeTainted
2020/10/05 18:02:46 [TRACE] eval: *terraform.EvalWriteState
2020/10/05 18:02:46 [TRACE] EvalWriteState: recording 2 dependencies for cloudfoundry_app.web_app
2020/10/05 18:02:46 [TRACE] EvalWriteState: writing current state object for cloudfoundry_app.web_app
2020/10/05 18:02:46 [TRACE] eval: *terraform.EvalIf
2020/10/05 18:02:46 [TRACE] eval: *terraform.EvalIf
2020/10/05 18:02:46 [TRACE] eval: *terraform.EvalWriteDiff
2020/10/05 18:02:46 [TRACE] eval: *terraform.EvalApplyPost
2020/10/05 18:02:46 [ERROR] eval: *terraform.EvalApplyPost, err: rpc error: code = Unavailable desc = transport is closing
2020/10/05 18:02:46 [ERROR] eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2020/10/05 18:02:46 [TRACE] [walkApply] Exiting eval tree: cloudfoundry_app.web_app
2020/10/05 18:02:46 [TRACE] vertex "cloudfoundry_app.web_app": visit complete
2020/10/05 18:02:46 [TRACE] dag/walk: upstream of "meta.count-boundary (EachMode fixup)" errored, so skipping
2020/10/05 18:02:46 [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/cloudfoundry-community/cloudfoundry\"] (close)" errored, so skipping
2020/10/05 18:02:46 [TRACE] dag/walk: upstream of "root" errored, so skipping
2020/10/05 18:02:46 [TRACE] statemgr.Filesystem: creating backup snapshot at terraform.tfstate.backup
2020/10/05 18:02:46 [TRACE] statemgr.Filesystem: state has changed since last snapshot, so incrementing serial to 29
2020/10/05 18:02:46 [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate
2020/10/05 18:02:46 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info
2020/10/05 18:02:46 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate using fcntl flock
2020-10-05T18:02:46.408+0100 [DEBUG] plugin: plugin exited
```

--- a/cloudfoundry/managers/appdeployers/bluegreenv2_strategy.go
+++ b/cloudfoundry/managers/appdeployers/bluegreenv2_strategy.go
@@ -147,6 +147,9 @@ func (s BlueGreenV2) Restage(appDeploy AppDeploy) (AppDeployResponse, error) {
 					BindTimeout:  appDeploy.BindTimeout,
 					StartTimeout: appDeploy.StartTimeout,
 				})
+				if err != nil {
+					return nil, err
+				}
 				ctx["app_response"] = AppDeployResponse{
 					App:             app,
 					RouteMapping:    rejoinMappingPort(app.Ports[0], appResp.RouteMapping),

--- a/docs/resources/app.md
+++ b/docs/resources/app.md
@@ -37,7 +37,7 @@ The following arguments are supported:
    * an empty blank string to use built-in buildpacks (i.e. autodetection)
 * `command` - (Optional, String) A custom start command for the application. This overrides the start command provided by the buildpack.
 * `enable_ssh` - (Optional, Boolean) Whether to enable or disable SSH access to the container. Default is `true` unless disabled globally.
-* `timeout` - (Optional, Number) Max wait time for app instance startup, in seconds
+* `timeout` - (Optional, Number) Max wait time for app instance startup, in seconds. Defaults to 60 seconds.
 * `stopped` - (Optional, Boolean) Defines the desired application state. Set to `true` to have the application remain in a stopped state. Default is `false`, i.e. application will be started.
 * `labels` - (Optional, map string of string) Add labels as described [here](https://docs.cloudfoundry.org/adminguide/metadata.html#-view-metadata-for-an-object). 
 Works only on cloud foundry with api >= v3.63.
@@ -145,6 +145,12 @@ The following attributes are exported along with any defaults for the inputs att
 * `id` - The GUID of the application
 * `id_bg` - The GUID of the application updated by resource when strategy is blue-green. 
 This allow change a resource linked to app resource id to be updated when app will be recreated.
+
+## Timeouts
+
+* App instance startup timeout - see the `timeout` argument.
+* App staging timeout - 15 mins.
+* Service binding timeout - 5 mins.
 
 ## Import
 

